src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
+                                       , _syntaxErrorLocations = mempty
+                                       })) env kont
+           (l:ls) -> Down (unCore l) env (InSyntaxErrorLocations msg_syn ls mempty env kont)
+          )
+      (\err -> Er err env kont)
+    -- done
+    (InSyntaxErrorLocations msg_syn [] dones env kont) ->
+        Up (ValueMacroAction
+                $ MacroActionSyntaxError (SyntaxError { _syntaxErrorMessage   = msg_syn
+                                                      , _syntaxErrorLocations = dones
+                                                      })) env kont
+    (InSyntaxErrorLocations msg (l:ls) dones env kont) ->
+      evalAsSyntax v
+      (\good -> Down (unCore l) env (InSyntaxErrorLocations msg ls (good : dones) env kont))
+      (\err  -> Er err env kont)
+
+-- the downsweep, searching for a redex to evaluate.
+step (Down c env k)  =
+  case c of
+
+    -- atoms
+    (CoreString s)    -> Up (ValueString s) env k
+    (CoreInteger i)   -> Up (ValueInteger i) env k
+    (CoreIntegerSyntax (ScopedInteger int scope)) -> Down (unCore int) env (InInteger scope env k)
+    (CoreStringSyntax  (ScopedString  str scope)) -> Down (unCore str) env (InString scope env k)
+    (CoreSyntax s)    -> Up (ValueSyntax s) env k
+    (CoreError what)  -> Down (unCore what) env (InError env k)
+    (CoreEmpty (ScopedEmpty scope)) -> Down (unCore scope) env (InScope (List mempty) env k)
+    CoreMakeIntroducer -> Up (ValueMacroAction MacroActionIntroducer)   env k
+    CoreWhichProblem   -> Up (ValueMacroAction MacroActionWhichProblem) env k
+
+
+    -- variables and binders
+    (CoreVar var) ->
+      case lookupVal var env of
+        Just val -> Up val env k
+        _        -> Er (EvalErrorUnbound var) env k
+
+    (CoreLet ident var def body) ->
+      Down (unCore def) env (InLetDef ident var (unCore body) env k)
+
+    (CoreLetFun fIdent fVar argIdent argVar def body) ->
+      let vFun = ValueClosure $ FO $ FOClosure
+            { _closureEnv   = Env.insert fVar fIdent vFun env
+            , _closureIdent = argIdent
+            , _closureVar   = argVar
+            , _closureBody  = def
+            }
+          newEnv = Env.insert fVar fIdent vFun env
+      in Down (unCore body) newEnv k
+
+    (CoreCtor con args) -> case args of
+                           -- just a symbol, shortcut out
+                           []     -> Up (ValueCtor con mempty) env k
+                           -- process fields left to right
+                           (f:fs) -> Down (unCore f) env (InCtor con (fmap unCore fs) mempty env k)
+
+
+    -- lambdas and application
+    (CoreLam ident var body) ->
+      let lam = ValueClosure $ FO $ FOClosure
+            { _closureEnv   = env
+            , _closureIdent = ident
+            , _closureVar   = var
+            , _closureBody  = body
+            }
+      in Up lam env k
+    (CoreApp fun arg) -> Down (unCore arg) env (InArg (unCore fun) env k)
```

Comment:
`eval` was recurring on the function position first, not the argument, we think the CEK version should do the same

(currently pair-reviewing with @david-christiansen)